RoomBookingSystem/app/Http/Controllers/BookingController.php

```diff
@@ -0,0 +1,125 @@
+                $query->whereBetween('start_time', [$validated['start_time'], $validated['end_time']])
+                    ->orWhereBetween('end_time', [$validated['start_time'], $validated['end_time']])
+                    ->orWhere(function ($query) use ($validated) {
+                        // Checks if the booking completely overlap another booking
+                        $query->where('start_time', '<=', $validated['start_time'])
+                            ->where('end_time', '>=', $validated['end_time']);
+                    });
+            })
+            ->exists();
+
+        if ($overlap) {
+            return response()->json(['message' => 'This time slot is already booked.'], 400);
+        }
+
+        $booking = Booking::create([
+            'title' => $validated['title'],
+            'description' => $validated['description'] ?? null,
+            'start_time' => $validated['start_time'],
+            'end_time' => $validated['end_time'],
+            'user_id' => $user->id,
+            'room_id' => $validated['room_id'],
+        ]);
+
+        return new BookingResource($booking);
+    }
+
+    public function updateBooking(Request $request, Booking $booking)
+    {
+        $user = auth()->user();
+        if ($user->role !== 'admin' && $user->id !== $booking->user_id) {
+            return response()->json(['message' => 'Unauthorized'], 401);
+        }
+
+        // Updates the model using only they valid values, not the datebase
+        $booking->fill($request->validate([
+            'title' => 'sometimes|string',
+            'description' => 'sometimes|string',
+            'start_time' => 'sometimes|date',
+            'end_time' => 'sometimes|date|after_or_equal:start_time',
+            'room_id' => 'sometimes|exists:rooms,id',
+        ]));
+
+        // Checks for overlap
+        $overlap = Booking::where('room_id', $booking->room_id)
+            ->where('id', '!=', $booking->id)
+            ->where(function ($query) use ($booking) {
+                $query->whereBetween('start_time', [$booking->start_time, $booking->end_time])
+                    ->orWhereBetween('end_time', [$booking->start_time, $booking->end_time])
+                    ->orWhere(function ($query) use ($booking) {
+                        $query->where('start_time', '<=', $booking->start_time)
+                            ->where('end_time', '>=', $booking->end_time);
+                    });
+            })
+            ->exists();
+
+        if ($overlap) {
+            return response()->json(['message' => 'This time slot is already booked.'], 400);
+        }
+
+        $booking->save();
+
+        return new BookingResource($booking);
+    }
+
+    public function delete(Booking $booking)
+    {
+        $user = auth()->user();
+        if ($user->role !== 'admin' && $user->id !== $booking->user_id) {
+            return response()->json(['message' => 'Unauthorized'], 401);
+        }
```

Comment:
Shouldn't it be forbidden instead, as the user is authenticated, but doesn't have permission to delete it.
```suggestion
        if ($user->role !== 'admin' && $user->id !== $booking->user_id) {
            return response()->json(['message' => 'Forbidden'], 403);
        }
```